File: retort-reflections/src/main/java/co/kaioru/retort/reflection/IReflectionGenerator.java
package co.kaioru.retort.reflection;

import co.kaioru.retort.ICommandContext;
import co.kaioru.retort.annotation.IAnnotationGenerator;

import java.util.Map;

public interface IReflectionGenerator<I extends ICommandContext, O> extends IAnnotationGenerator<I, O> {

    Map<Class<?>, IReflectionProvider<I, ?>> getProviders();

    default void registerProvider(IReflectionProvider<I, ?> provider) {
        getProviders().put(provider.getType(), provider);
    }

    default void deregisterProvider(IReflectionProvider<I, ?> provider) {
        getProviders().remove(provider.getType());
    }

}

File: retort-reflections/src/main/kotlin/co/kaioru/retort/reflection/IReflectionGenerator.kt
package co.kaioru.retort.reflection

import co.kaioru.retort.annotation.IAnnotationGenerator

interface IReflectionGenerator<I, O : Any> : IAnnotationGenerator<I, O> {
    val providers: MutableMap<Class<*>, IReflectionProvider<I, *>>

    fun registerProvider(provider: IReflectionProvider<I, *>) {
        this.providers.put(provider.type, provider)
    }

    fun registerProviders(providers: Collection<IReflectionProvider<I, *>>) {
        providers.forEach { registerProvider(it) }
    }
}